File: tests/fixtures/rules-tests/knockout/knockout.ts
// KNOCKOUT: completely unrelated code — no LLM, no security context
// This file should NOT trigger any security rule.

function deepClone<T>(obj: T): T {
    return JSON.parse(JSON.stringify(obj));
}

function chunkArray<T>(arr: T[], size: number): T[][] {
    const result: T[][] = [];
    for (let i = 0; i < arr.length; i += size) {
        result.push(arr.slice(i, i + size));
    }
    return result;
}

function debounce<T extends (...args: any[]) => any>(
    fn: T,
    delay: number
): (...args: Parameters<T>) => void {
    let timeoutId: ReturnType<typeof setTimeout>;
    return (...args: Parameters<T>) => {
        clearTimeout(timeoutId);
        timeoutId = setTimeout(() => fn(...args), delay);
    };
}

function groupBy<T>(items: T[], keyFn: (item: T) => string): Record<string, T[]> {
    return items.reduce((acc, item) => {
        const key = keyFn(item);
        if (!acc[key]) acc[key] = [];
        acc[key].push(item);
        return acc;
    }, {} as Record<string, T[]>);
}

File: tests/fixtures/rules-tests/knockout/knockout.cs
// KNOCKOUT: completely unrelated code — no security context, no NotImplementedException
// This file should NOT trigger any security rule.
using System;
using System.Collections.Generic;
using System.Linq;

public static class MathHelpers
{
    public static long Factorial(int n)
    {
        if (n <= 1) return 1;
        long result = 1;
        for (int i = 2; i <= n; i++) result *= i;
        return result;
    }

    public static double Median(double[] values)
    {
        Array.Sort(values);
        int n = values.Length;
        if (n % 2 == 0)
            return (values[n / 2 - 1] + values[n / 2]) / 2.0;
        return values[n / 2];
    }

    public static Dictionary<T, int> CountFrequency<T>(IEnumerable<T> items) where T : notnull
    {
        return items.GroupBy(x => x).ToDictionary(g => g.Key, g => g.Count());
    }

    public static string Reverse(string s)
    {
        return new string(s.Reverse().ToArray());
    }
}
